### PR TITLE
[18.0][MIG] product_template_multi_link: Migration to 18.0

### PR DESCRIPTION
**Pull request by thienvh332**
### Blacklist:
- `#800`

### Changes:
- Delete name_search. Because it is no longer needed. (see https://github.com/odoo/odoo/pull/174967/commits/9e21ec27efb2dd2bcbb2d8fc1e0e2bad6ca8d62d#diff-35eeb9b2264b76eca30fcb0a923757dde3ca14c636ce6faf66ef31a6b649799bR539)

### CODE REVIEW

File: product_template_multi_link/models/product_template.py [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+            link_dict.setdefault(link_id, 0)
+            link_dict[link_id] += dic["right_product_tmpl_id_count"]
+
+        for rec in self:
+            rec.product_template_link_count = link_dict.get(rec.id, 0)
+
+    def show_product_template_links(self):
+        self.ensure_one()
+        return {
+            "name": _("Product links"),
```
Comment (nguyenminhchien) [inline product_template_multi_link/models/product_template.py]:
```suggestion
            "name": self.env._("Product links"),
```

File: product_template_multi_link/models/product_template_link.py [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        Check whether:
+            - the two products are different
+            - there is only one link between the same two templates for the same type
+
+        :raise: ValidationError if not ok
+        """
+        self.flush_recordset()  # flush required since the method uses plain sql
+        if any(rec._check_product_not_different() for rec in self):
+            raise ValidationError(
+                _("You can only create a link between 2 different products")
```
Comment (nguyenminhchien) [inline product_template_multi_link/models/product_template_link.py]:
```suggestion
                self.env._("You can only create a link between 2 different products")
```

File: product_template_multi_link/models/product_template_link.py [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+        self.env.cr.execute(query, query_args)
+        res = self.env.cr.fetchall()
+        is_duplicate_by_link_id = dict(res)
+        if True in is_duplicate_by_link_id.values():
+            ids = [k for k, v in is_duplicate_by_link_id.items() if v]
+            descrs = "\n ".join(
+                [link._duplicate_link_error_msg() for link in self.browse(ids)]
+            )
+            raise ValidationError(
+                _(
```
Comment (nguyenminhchien) [inline product_template_multi_link/models/product_template_link.py]:
```suggestion
                self.env._(
```

File: product_template_multi_link/models/product_template.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+    def get_links_by_code(self, code):
+        """Get all active active links maching code for current product."""
+        self.ensure_one()
+        return self.product_template_link_ids.filtered(
+            lambda r: r.type_id.code == code and r.is_link_active
+        )
+
+    @api.model
+    def _search_display_name(self, operator, value):
```
Comment (nguyenminhchien) [inline product_template_multi_link/models/product_template.py]: no need to override this method.
def _search_display_name() is different from def name_search().

Remove the context `name_search_default_code` from the view as well.

File: product_template_multi_link/models/product_template_link.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    )
+    right_product_tmpl_id = fields.Many2one(
+        string="Linked Product",
+        comodel_name="product.template",
+        required=True,
+        ondelete="cascade",
+        index=True,
+    )
+    right_product_img = fields.Image(
+        string="Right Product Image", related="right_product_tmpl_id.image_128"
```
Comment (nguyenminhchien) [inline product_template_multi_link/models/product_template_link.py:35]:
```
2024-12-04 07:07:57,750 148504 WARNING v18c_product_template_multi_link odoo.addons.base.models.ir_model: Two fields (right_product_img, left_product_img) of product.template.link() have the same label: Image 128. [Modules: product_template_multi_link and product_template_multi_link]
```

File: product_template_multi_link/models/product_template_link.py
Comment (nguyenminhchien) [inline product_template_multi_link/models/product_template_link.py:1]:
```
2024-12-04 08:21:19,556 148504 INFO v18c_product_template_multi_link werkzeug: 127.0.0.1 - - [04/Dec/2024 08:21:19] "POST /web/dataset/call_kw/product.template.link/onchange HTTP/1.1" 200 - 2 0.002 0.007
2024-12-04 08:21:20,172 148504 ERROR v18c_product_template_multi_link odoo.osv.expression: Non-stored field product.template.link.type.display_name cannot be searched. 
NoneType: None
```

File: product_template_multi_link/views/product_template_link_view.xml [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+                    context="{'name_search_default_code': True}"
+                />
+            </list>
+        </field>
+    </record>
+    <record id="product_template_link_kanban" model="ir.ui.view">
+        <field name="name">Product.template.link.kanban</field>
+        <field name="model">product.template.link</field>
+        <field name="arch" type="xml">
+            <kanban>
```
Comment (nguyenminhchien) [inline product_template_multi_link/views/product_template_link_view.xml:44]: Kanban view is broken.
image

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]

**Comment by thienvh332** [comment]
Hi @nguyenminhchien ,
- I'm not sure what's going on your side. The errors you had I fixed a long time ago and I don't see them locally. If it happens again let me know.
> Remove the context name_search_default_code from the view as well.

- In my opinion, to avoid affecting the search results of other flows, we still need context here. What do you think?

**Comment by nguyenminhchien** [comment]
> Hi @nguyenminhchien ,
> 
>     * I'm not sure what's going on your side. The errors you had I fixed a long time ago and I don't see them locally. If it happens again let me know.
> 
> 
> > Remove the context name_search_default_code from the view as well.
> 
>     * In my opinion, to avoid affecting the search results of other flows, we still need context here. What do you think?

Could you please give me a test case which triggers this search condition? @thienvh332 
```
        if name and len(args or []) == 1 and args[0][0] == "id" and search_default_code:
            args = expression.AND(
                [
                    args,
                    expression.OR(
                        [
                            [("default_code", operator, name)],
                            [("product_variant_ids.default_code", operator, name)],
                            [(self._rec_name, operator, name)],
                        ]
                    ),
                ]
            )
            # Reset 'name' so base '_name_search' won't add '_rec_name'
            # to 'args' (already added above).
            # See 'odoo.models.BaseModel._name_search'.
            name = ""
```

**Comment by thienvh332** [comment]
Hi @nguyenminhchien : [Link testcase](https://github.com/thienvh332/e-commerce/blob/18.0-mig-product_template_multi_link/product_template_multi_link/tests/test_product_template.py#L40)

**Comment by nguyenminhchien** [comment]
> Hi @nguyenminhchien : [Link testcase](https://github.com/thienvh332/e-commerce/blob/18.0-mig-product_template_multi_link/product_template_multi_link/tests/test_product_template.py#L40)

Let me correct myself, i means "user case" not "test case".
I want a user case in 18.0 which triggers it.

**Comment by thienvh332** [comment]
We cannot have this case in UI because **`name_search`** will not pass condition **`args[0][0] == "id"`** which will make "if" false and this case will not be triggered
image

**Comment by thienvh332** [comment]
We can consider allowing search with `default_code` without `id != 0` condition when there is context `name_search_default_code`. What do you think?

**Review by nguyenminhchien — CHANGES_REQUESTED** [review]
If you remove name_search.
You should not port the commit ` pt_multi_link: improve UX by searching templates on default_code `. But, add it to ignore list, instead.

**Review by nguyenminhchien — APPROVED** [review]

**Pull request closed without merging**